### PR TITLE
Move check_label ci to mergebot

### LINKED ISSUE #88098: [Label Check] Move label check logic from CI to mergebot
Currently, we have [a CI check](https://github.com/pytorch/pytorch/blob/master/.github/workflows/check-labels.yml#L44) that checks to see if a PR has the appropriate labels on it, and fails the check if the right label is missing.

Instead of making this be something that makes the CI appear to have failed, lets make two changes:

1. Modify the CI check to only share a comment reminding the user to add the appropriate CI (without reporting a failing check)
2. Update mergebot rules in https://github.com/pytorch/pytorch/blob/master/.github/scripts/trymerge.py to 1) fail the merge if the PR is missing the required labels and 2) post a comment explaining that the merge was failed due to the missing label

Original customer request:

> Can we make the "needs labels" signal not part of the actual CI signal? I use the GitHub green/red indicator to easily check if a PR succeeded CI or not, but lack of labels turns this into a fail and obscures the signal.
> 
> Instead, **the merge bot should just reject merges if they don't have a label**. The idea being, it's very easy to immediately add the label and try again. You don't need this signal ahead of time!

### PR DESCRIPTION
**Pull request by Goldspear**
Fixes #88098

### What Changed
* Moved `check_label.py` logic into `trymerge.py`
* Refactored relevant unittests
* ~~Dropped~~ Refactored `check_label.py` ci job

### Tests
`python .github/scripts/test_trymerge.py`
`python .github/scripts/test_check_labels.py`
`make lint & lintrunner -a`

### Notes to reviewers
This PR replaces the [original PR](https://github.com/pytorch/pytorch/pull/92225) to workaround the sticky EasyCLA failure mark on its first commit.

### CODE REVIEW

**Comment by Goldspear** [comment]
cc @ZainRizvi for first-time contributor approval and review.

File: .github/scripts/trymerge.py [hunk 1]
```diff
@@ -1511,6 +1548,12 @@ def handle_exception(e: Exception, title: str = "Merge failed") -> None:
         gh_post_pr_comment(org, project, args.pr_num, "Cross-repo ghstack merges are not supported", dry_run=args.dry_run)
         return
 
+    if not has_required_labels(pr):
+        add_label_err_comment(pr)
+        return
+    else:
+        delete_all_label_err_comments(pr)
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:1555]: Within trymerge, the failure here isn't as loud as we'd like it to be. The `handle_exception` method formats it in a better way and makes the message easier to parse.

Can you please move this checking/commenting to be inside the `merge` method and to use the error formatting over there?

**Review by ZainRizvi — DISMISSED** [review]

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
This is looking pretty close, just a couple last pieces left! (see prev comment)

File: .github/scripts/trymerge.py (hunk 1 above)
Comment (Goldspear) [reply .github/scripts/trymerge.py]: Ah good point! Moved into merge to take advantage of existing error handling.

File: .github/scripts/trymerge.py [hunk 2]
```diff
@@ -413,6 +414,16 @@ class WorkflowCheckState(NamedTuple):
...
 
+BOT_AUTHORS = ["github-actions", "pytorchmergebot", "pytorch-bot"]
+
+LABEL_ERR_MSG_TITLE = "This PR needs a label"
+LABEL_ERR_MSG = (
+    f"# {LABEL_ERR_MSG_TITLE}\n"
+    "If your changes are user facing and intended to be a part of release notes, please use a label starting with `release notes:`.\n\n"  # noqa: E501  pylint: disable=line-too-long
+    "If not, please add the `topic: not user facing` label.\n\n"
+    "For more information, see https://github.com/pytorch/pytorch/wiki/PyTorch-AutoLabel-Bot#why-categorize-for-release-notes-and-how-does-it-work."  # noqa: E501  pylint: disable=line-too-long
+)
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:425]: Thinking about how the formatting will look when there's a merge failed, with the current approach we'll end up with two H1 heading lines back to back (here's [an example](https://github.com/pytorch/pytorch/pull/92107#issuecomment-1382241578) of how the comment looks).  To fix that, the "This PR needs a label" line shouldn't be treated as a header when it's failing a merge

We could take something like the following and stick `MERGE_LABEL_ERR_MSG` in the merge failure message

```suggestion
LABEL_ERR_MSG_TITLE = "This PR needs a label"
LABEL_ERR_MSG_BODY = (
    "If your changes are user facing and intended to be a part of release notes, please use a label starting with `release notes:`.\n\n"  # noqa: E501  pylint: disable=line-too-long
    "If not, please add the `topic: not user facing` label.\n\n"
    "For more information, see https://github.com/pytorch/pytorch/wiki/PyTorch-AutoLabel-Bot#why-categorize-for-release-notes-and-how-does-it-work."  # noqa: E501  pylint: disable=line-too-long
)
LABEL_ERR_MSG = (
    f"# {LABEL_ERR_MSG_TITLE}\n"
    f"{LABEL_ERR_MSG_BODY}"
)
MERGE_LABEL_ERR_MSG = (
    f"{LABEL_ERR_MSG_TITLE}\n"
    f"{LABEL_ERR_MSG_BODY}"
)
```

File: .github/scripts/trymerge.py [hunk 3]
```diff
@@ -1390,6 +1427,11 @@ def merge(pr_num: int, repo: GitRepo,
     # here to stop the merge process right away
     find_matching_merge_rule(pr, repo, skip_mandatory_checks=True)
 
+    if not has_required_labels(pr):
+        raise RuntimeError(LABEL_ERR_MSG)
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:1431]: Then this bit would change to:

```suggestion
        raise RuntimeError(MERGE_LABEL_ERR_MSG)
```

File: .github/workflows/check-labels.yml [hunk 4]
```diff
@@ -29,12 +29,10 @@ jobs:
           cache: pip
           cache-dependency-path: |
             **/.github/requirements-gha-cache.txt
-
```
Comment (ZainRizvi) [inline .github/workflows/check-labels.yml:32]: Were these two line break deletions accidental?

File: .github/scripts/test_trymerge.py [hunk 5]
```diff
@@ -347,6 +433,7 @@ def test_main_merge(self, mock_merge: Any, mock_parse_args: Any, mock_gh_get_inf
                                            land_checks=False,
                                            mandatory_only=False)
 
+
```
Comment (ZainRizvi) [inline .github/scripts/test_trymerge.py:436]: nit: extra newline

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
Getting very close :D

File: .github/scripts/trymerge.py (hunk 2 above)
Comment (Goldspear) [reply .github/scripts/trymerge.py:425]: Ah, great catch and thanks for sharing the comment example! I was wondering if there's a way to actually visualise the bot comments on GH.

File: .github/workflows/check-labels.yml (hunk 4 above)
Comment (Goldspear) [reply .github/workflows/check-labels.yml:32]: I thought these might be lintrunner autofix, but let me directly checkout master in this case.

File: .github/scripts/trymerge.py (hunk 2 above)
Comment (ZainRizvi) [reply .github/scripts/trymerge.py:425]: The way I do it, I find a random merge comment on github and then apply the edits my change *should* make and see how it ends up looking

**Review by ZainRizvi — APPROVED** [review]
Nice!

**Comment by Goldspear** [comment]
Hi @ZainRizvi, it seems the three ci-tests failed again last night at `Upload Test Statistics` step with a 
```
ModuleNotFoundError: No module named 'typing_extensions'
Error: Process completed with exit code 1.
```
I've just done a rebase to resolve some conflicts with the new master, so will need your blessing again to trigger the merge ci jobs :)

**Comment by malfet** [comment]
Had to revert it manually, as it broke bot merge and revert bot workflows, see https://github.com/pytorch/pytorch/actions/runs/3963082664 and https://github.com/pytorch/pytorch/actions/runs/3963235531/jobs/6790838677

**Comment by malfet** [comment]
@ZainRizvi please loop me in the followup revision of this change?

**Pull request closed without merging**